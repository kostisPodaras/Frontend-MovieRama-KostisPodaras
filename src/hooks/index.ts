export * from './useGenres';
export * from './useMovies';
export * from './useSearchMovies';
export * from './useMovieDetails';
