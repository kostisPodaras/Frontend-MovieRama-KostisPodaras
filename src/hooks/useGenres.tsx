import { useEffect, useState } from 'react';
import { API_GENRES } from '../API';

interface Genre {
  id: number;
  name: string;
}

export const useGenres = (): {
  genres: Genre[];
  genresIsError: boolean;
  genresIsLoading: boolean;
} => {
  const [genres, setGenres] = useState<Genre[]>([]);
  const [genresIsError, setGenresIsError] = useState(false);
  const [genresIsLoading, setGenresIsLoading] = useState(false);

  useEffect(() => {
    try {
      setGenresIsLoading(true);

      const fetchGenres = async () => {
        const response = await fetch(API_GENRES);
        const data = await response.json();

        setGenres(data.genres);
        setGenresIsLoading(false);
      };
      fetchGenres();
    } catch (error) {
      setGenresIsError(true);
      console.warn(error);
    }
  }, []);

  return {
    genres,
    genresIsLoading,
    genresIsError,
  };
};
