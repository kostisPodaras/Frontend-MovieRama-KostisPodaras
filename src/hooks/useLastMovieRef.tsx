import { useRef, useCallback } from 'react';

export const useLastMovieRef = (isLoading, hasNextPage, setPageNumber) => {
  const intersectionObserver = useRef<IntersectionObserver | null>(null);
  const lastMovieRef = useCallback(
    (movie: HTMLDivElement) => {
      if (isLoading) return;

      if (intersectionObserver.current) {
        intersectionObserver.current.disconnect();
      }

      intersectionObserver.current = new IntersectionObserver((movies) => {
        if (movies[0].isIntersecting && hasNextPage) {
          setPageNumber((prev) => prev + 1);
        }
      });

      if (movie) intersectionObserver.current.observe(movie);
    },
    [isLoading, hasNextPage],
  );

  return lastMovieRef;
};
