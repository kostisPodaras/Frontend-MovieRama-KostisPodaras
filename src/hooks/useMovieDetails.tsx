import { useEffect, useState } from 'react';

import { Review, Movie, Similar } from 'types';

import { getMovieDetails } from '../services';

interface Video {
  id: number;
  key: string;
  name: string;
  site: string;
  type: string;
  official: boolean;
}

interface Response extends Movie {
  videos: {
    results: Video[];
  };
  reviews: {
    results: Review[];
  };
  similar: {
    results: Similar[];
  };
}

export const useMovieDetails = (movieId: number) => {
  const [movieDetails, setMovieDetails] = useState<Response>();
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    getMovieDetails(movieId)
      .then((data) => {
        setMovieDetails(data);
        setIsLoading(false);
      })
      .catch((error) => {
        console.warn(error);
        setIsLoading(false);
        setIsError(true);
      });
  }, []);

  return {
    movieDetails,
    isLoading,
    isError,
  };
};
