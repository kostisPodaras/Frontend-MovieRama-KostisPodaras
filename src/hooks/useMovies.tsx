import { useEffect, useState } from 'react';
import { MovieProps } from '../components/movies-list/MoviesList';
import { getUniqueElements } from '../utils';
import { getMovies, searchMovies } from '../models';

const useMovies = (pageNumber = 1, query: string) => {
  const [results, setResults] = useState<MovieProps[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);
  const [hasNextPage, setHasNextPage] = useState(false);

  useEffect(() => {
    // Condition so we dont make requests for now_playing_movies when we are on search and pageNumber change
    if (!query) {
      setIsLoading(true);
      setIsError(false);

      getMovies(pageNumber)
        .then(({ data, isLastPage }) => {
          setResults((prev: MovieProps[]) => {
            // Filtering out duplicate movies, seems pages from API contain duplicate movies
            const uniqueMovies = getUniqueElements([...prev, ...data], 'id');

            return uniqueMovies;
          });
          setHasNextPage(!isLastPage);
          setIsLoading(false);
        })
        .catch((error) => {
          console.warn(error);
          setIsLoading(false);
          setIsError(true);
        });
    }
  }, [pageNumber]);

  return {
    theaterMovies: results,
    theaterMoviesIsLoading: isLoading,
    theaterMoviesIsError: isError,
    theaterMoviesHasNextPage: hasNextPage,
  };
};

export default useMovies;
