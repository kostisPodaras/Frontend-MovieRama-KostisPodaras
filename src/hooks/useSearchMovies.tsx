import { useEffect, useState } from 'react';
import { MovieProps } from '../components/movies-list/MoviesList';
import { getUniqueElements } from '../utils';
import { searchMovies } from '../services';

export const useSearchMovies = (pageNumber = 1, query: string) => {
  const [queryResults, setQueryResults] = useState<MovieProps[]>([]);
  const [queryIsError, setQueryIsError] = useState(false);
  const [queryIsLoading, setQueryIsLoading] = useState(false);
  const [queryHasNextPage, setQueryHasNextPage] = useState(false);
  const [prevQuery, setPrevQuery] = useState(query);

  useEffect(() => {
    if (query) {
      setQueryIsLoading(true);
      setQueryIsError(false);

      const isQueryChanged = query !== prevQuery;

      const debounce = setTimeout(() => {
        // When we type again, we reset the pageNumber. (Bug was If we scroll to bottom and go to page 2, then scrolling back up and typing, will search th results of page 2)
        searchMovies(isQueryChanged ? 1 : pageNumber, query)
          .then(({ data, isLastPage }) => {
            setQueryResults((prev: MovieProps[]) => {
              // Filtering out duplicate movies, seems pages from API contain duplicate movies
              const uniqueMovies =
                // If query change, show only the new movies without the prev movies
                isQueryChanged
                  ? getUniqueElements(data, 'id')
                  : getUniqueElements([...prev, ...data], 'id');

              return uniqueMovies;
            });
            setPrevQuery(query);
            setQueryHasNextPage(!isLastPage);
            setQueryIsLoading(false);
          })
          .catch((error) => {
            console.warn(error);
            setQueryIsLoading(false);
            setQueryIsError(true);
          });
      }, 800);

      return () => {
        clearTimeout(debounce);
      };
    }
  }, [pageNumber, query]);

  return {
    queryResults,
    queryIsLoading,
    queryIsError,
    queryHasNextPage,
  };
};
