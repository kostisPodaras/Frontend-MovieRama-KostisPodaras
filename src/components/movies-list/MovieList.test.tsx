import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import * as hooks from 'hooks/useGenres';

import { MoviesList } from '..';

describe('MovieList component', () => {
  const mockProps = [
    {
      genre_ids: [28, 14, 878],
      id: 436270,
      overview:
        'Nearly 5,000 years after he was bestowed with the almighty powers of the Egyptian gods—and imprisoned just as quickly—Black Adam is freed from his earthly tomb, ready to unleash his unique form of justice on the modern world.',
      poster_path: '/pFlaoHTZeyNkG83vxsAJiGzfSsa.jpg',
      release_date: '2022-10-19',
      title: 'Black Adam',
      vote_average: 7.3,
    },
    {
      genre_ids: [28, 18, 36],
      id: 724495,
      overview:
        'The story of the Agojie, the all-female unit of warriors who protected the African Kingdom of Dahomey in the 1800s with skills and a fierceness unlike anything the world has ever seen, and General Nanisca as she trains the next generation of recruits and readies them for battle against an enemy determined to destroy their way of life.',
      poster_path: '/lQMJHnHxUyj8kJlC2YOKNuzuwMP.jpg',
      release_date: '2022-09-15',
      title: 'The Woman King',
      vote_average: 7.9,
    },
  ];

  test('Should render the component', () => {
    jest.spyOn(hooks, 'useGenres').mockImplementation(() => {
      return {
        genres: [
          { id: 28, name: 'Action' },
          { id: 12, name: 'Adventure' },
          { id: 16, name: 'Animation' },
        ],
        genresIsLoading: false,
        genresIsError: false,
      };
    });

    render(<MoviesList movies={mockProps} lastMovieRef={null} />);
    const mainElement = screen.getByRole('main');
    expect(mainElement).toBeInTheDocument();
  });

  test('Should render all the chips plus the "all" chip', () => {
    jest.spyOn(hooks, 'useGenres').mockImplementation(() => {
      return {
        genres: [
          { id: 28, name: 'Action' },
          { id: 12, name: 'Adventure' },
          { id: 16, name: 'Animation' },
        ],
        genresIsLoading: false,
        genresIsError: false,
      };
    });

    render(<MoviesList movies={[]} lastMovieRef={null} />);
    const chipElement = screen.getAllByTestId('chip');
    expect(chipElement).toHaveLength(4);
  });

  test('matches snapshot', () => {
    jest.spyOn(hooks, 'useGenres').mockImplementation(() => {
      return {
        genres: [
          { id: 28, name: 'Action' },
          { id: 12, name: 'Adventure' },
          { id: 16, name: 'Animation' },
        ],
        genresIsLoading: false,
        genresIsError: false,
      };
    });

    const snapshot = renderer.create(
      <MoviesList movies={mockProps} lastMovieRef={null} />,
    );
    expect(snapshot).toMatchSnapshot();
  });
});
