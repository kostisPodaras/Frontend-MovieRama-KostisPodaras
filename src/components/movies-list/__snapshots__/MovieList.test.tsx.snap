// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component matches snapshot 1`] = `
<main>
  <div
    className="container chipsOverride"
  >
    <div
      className="chip active"
      data-testid="chip"
      onClick={[Function]}
    >
      <p
        className="chipText"
      >
        all
      </p>
    </div>
    <div
      className="chip "
      data-testid="chip"
      onClick={[Function]}
    >
      <p
        className="chipText"
      >
        Action
      </p>
    </div>
    <div
      className="chip "
      data-testid="chip"
      onClick={[Function]}
    >
      <p
        className="chipText"
      >
        Adventure
      </p>
    </div>
    <div
      className="chip "
      data-testid="chip"
      onClick={[Function]}
    >
      <p
        className="chipText"
      >
        Animation
      </p>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      data-testid="card"
    >
      <div
        className="container"
        data-testid="card-background"
        onClick={[Function]}
        style={
          Object {
            "backgroundImage": "url(https://image.tmdb.org/t/p/w500/pFlaoHTZeyNkG83vxsAJiGzfSsa.jpg)",
          }
        }
      >
        <div
          className="genresWrapper"
        >
          <div
            className="container undefined"
          >
            <div
              className="chip "
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              >
                Action
              </p>
            </div>
            <div
              className="chip active"
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              />
            </div>
            <div
              className="chip active"
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              />
            </div>
          </div>
        </div>
        <div
          className="overviewOverlay"
        >
          <p>
            Nearly 5,000 years after he was bestowed with the almighty powers of the Egyptian gods—and imprisoned just as quickly—Black Adam is freed from his earthly tomb, ready to unleash his unique form of justice on the modern world.
          </p>
        </div>
      </div>
      <p
        className="text"
      >
        Black Adam
         | 
        2022
      </p>
      <div
        className="rating"
      >
        <div
          className="aboveAverage rating"
        />
        <p>
          7.3
        </p>
      </div>
    </div>
    <div
      data-testid="card"
    >
      <div
        className="container"
        data-testid="card-background"
        onClick={[Function]}
        style={
          Object {
            "backgroundImage": "url(https://image.tmdb.org/t/p/w500/lQMJHnHxUyj8kJlC2YOKNuzuwMP.jpg)",
          }
        }
      >
        <div
          className="genresWrapper"
        >
          <div
            className="container undefined"
          >
            <div
              className="chip "
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              >
                Action
              </p>
            </div>
            <div
              className="chip active"
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              />
            </div>
            <div
              className="chip active"
              data-testid="chip"
              onClick={[Function]}
            >
              <p
                className="chipText"
              />
            </div>
          </div>
        </div>
        <div
          className="overviewOverlay"
        >
          <p>
            The story of the Agojie, the all-female unit of warriors who protected the African Kingdom of Dahomey in the 1800s with skills and a fierceness unlike anything the world has ever seen, and General Nanisca as she trains the next generation of recruits and readies them for battle against an enemy determined to destroy their way of life.
          </p>
        </div>
      </div>
      <p
        className="text"
      >
        The Woman King
         | 
        2022
      </p>
      <div
        className="rating"
      >
        <div
          className="aboveAverage rating"
        />
        <p>
          7.9
        </p>
      </div>
    </div>
  </div>
</main>
`;
